### ISSUE TITLE
Improve SAS macros by allowing CSV files of included items

### ISSUE THREAD

**Issue by depryf** [imsweb/naaccr-xml#197]
As more people try to use the macros, I receive more feedback.

One thing that seems common is to handle multiple files with a single call to the macros.

The use case is to process all the files in a given folder, read them in a single data set, and process that unique data set (computing some values/stats and outputting them for example).

The "source data file" parameter of the read macro could be improved. Right now it is the full path or relative path of a unique source data file.  It would be simple to allow multiple files using a separator (a pipe for example).

An even more powerful improvement would be to recognize that the provided source is a folder instead of a file, and to process all the files in that folder. For this feature to properly work, it would also need some kind of regular expression that is executed on each filename so that it's only executed if it follows that regex. That's useful to ignore other file formats for example.

The other improvement would be for the list of items. It would be nice to allow the list to be provided as a CSV file for example.

**Comment by tuyetthieu2021** [comment]
Hi Fabian,
I am using the SAS Macro v7.8 to create XML (using write macro).  The output XML file is passed GenEdit Plus but it could not be recognized by NPCR Inter-Record Edit Tool version 5.2.6.  The error message said "The file does not appear to be in NAACCR version 18 format.

I use the NAACCR XML Utility v7.8 to convert the XML file (from SAS) to a new XML file with option #3 checked: "When writing the items in the XML file, also include the NAACCR Number."  The new output XML file from NAACCR XML Utility is accepted and processed by the Inter-Record Edit tool from record 1 until the last patient record with message: "Error 5 Description: Value of '1984' is not valid for "Value'.  'Value' should be between 'minimum' and 'maximum'. Parameter name: Value".

The total patient number in XML file should be 1,982 with 2,226 tumors records. But for some unknown reason, the Inter-Record Edit tool processed up to 1,983 or 1,984 which is causing the error.  I attached the screenshot so if you can see where is the issue.

**Comment by tuyetthieu2021** [comment]
Sorry, I forgot to mention that I am using base dictionary only, no user defined dictionary.

**Comment by depryf** [comment]
Hi @tuyetthieu2021 

So it looks like we are talking about two different problems.

1. The write SAS macro doesn't write NAACCR numbers and it looks like some NPCR software require them. It is a bit weird those software need the numbers since they are defined as optional in the NAACCR XML specifications. But regardless, I don't think it would be difficult to add an option to the macro to output the numbers. I will move that to its own issue.

2. You managed to run your file through the Inter-Record Edit tool but it seems to process an extra patient at the end of the file. It is difficult to know if this is an issue with the Inter-Record Edit tool or the XML data file (and the error message is really not helping). Can you try to re-open that file with the NAACCR XML Utility, but please use the "XML Validation" option (one of the buttons on the left toolbar). And then please process the file; that will run a full validation on it. If that tool says the file is valid, then this is most likely an issue with the Inter-Record Edit and I would try to contact the NPCR tech support.

**Comment by tuyetthieu2021** [comment]
Hi Fabian,

Thank you for quick response!  Yes, I did try to validation XML using NAACCR XML Utility and there is no warning.  Do you know who I can contact from NPCR tech support?

**Comment by depryf** [comment]
OK, I think it would be worth ask NPCR about this. 

They so have a tech support email on their website

https://www.cdc.gov/cancer/npcr/tools/edits/edittech.htm

**Comment by depryf** [comment]
Another improvement to the read/write macro that was suggested is to allow a CSV file for the list of included items, instead of having to specify the item XML IDs within the SAS code. That should be simple enough.

**Comment by depryf** [comment]
I feel like processing all the files in a given folder is a bit outside the scope of this library. That type of process can be setup with shell or batch scripts to call a SAS process multiple times. 

If there is a very strong demand for this, I will re-consider. 

For now I think I will only add support for providing a CSV file for the included data items.

**Comment by depryf** [comment]
So it's simple enough to allow a CSV file to be provided instead of a list of items. But a CSV file can have multiple columns. It can also have optional headers.

There are two ways to resolve that:

1. Force a single convention and hard-code it (the items must be defined in the first column of the CSV; they must always be a header.

2. Allow the users to provide that information along with the file. 

I am on the fence. Allowing the information to be provided is much more flexible, but it does add some complexity to this issue.

Current syntax:
```
items="patientIdNumber,primarySite"
```

New syntax:
```
items="included-items.csv",
items="C:\included-items.csv",
```

I need to think more about the headers and column indexes.

**Comment by depryf** [comment]
It's trivial to open those CSV files in Excel and add a header or move columns around. 

I think I am going to keep things simple for now and just assume that the first column is the one containing the XML IDs, and the first row is a header.  That is consistent with the format of the CSV dictionary that the SAS macros use. So a user will be able to open a "full" CSV dictionary in Excel, apply some filters to remove rows, and save the file in the same format, and use it for this parameters. I think that will be enough for this feature.

**Comment by depryf** [comment]
The work is done, but I still need to update the help in the macros and on the wiki. I will wait until I release the library so the dates in the changelogs make more sense.  I will keep this issue opened for now.

**Comment by depryf** [comment]
This is going to be released with version 7.12 of the library.
